### DISCUSSION TITLE
How does one debug a failed transaction on zksync

### DISCUSSION THREAD

**Discussion by cryptoman04** [zkSync-Community-Hub/zksync-developers#183] [General] [answered]
I'm building a dapp where transactions don't always work. To specify I'm calling a method on a contract and sometimes it fails. The problem is that I have no way to really understand and debug the failure. Is there some sort of stack trace I can look at? 

What are some tools I can use to better understand why some transactions fail on zksync. On Ethereum I was using tenderly, there doesn't seem to be something similar for zksync.

**Comment by dutterbutter** [comment, marked as answer]
Hey 👋 

Have you tried out era-test-node? It provides the ability to replay tx and give deep insights into why transactions fail. Here are some relevant links:
- [Docs](https://era.zksync.io/docs/tools/testing/era-test-node.html)
- [Repo](https://github.com/matter-labs/era-test-node)
- [Release Videos](https://www.youtube.com/playlist?list=PL-Bw34ynlUOMKWaT1yuXIvJ0E56Zji1cJ)

**Comment by patelkaren** [comment]
Unfortunately, ZKSync does not provide a direct equivalent to tools like Tenderly for Ethereum transaction debugging. However, there are several approaches you can take to better understand why transactions fail on ZKSync:

Check Transaction Status: After submitting a transaction, monitor its status on the ZKSync Explorer or through programmatic queries using the ZKSync API. This can give you insights into whether the transaction was successfully executed or if it failed.

Debugging Logs: If you have access to the contract code, ensure that it emits appropriate events or logs when transactions fail. You can then retrieve these logs using the ZKSync API or directly from the contract events.

Transaction Reverting: Transactions on ZKSync can revert for various reasons, including insufficient funds, invalid parameters, or errors in the contract code. Review your contract logic and error handling mechanisms to identify potential issues.

Contract Interactions: Verify that your contract interactions adhere to the requirements and constraints imposed by ZKSync, such as supported operations, gas limits, and supported data types.

Community Support: Engage with the ZKSync community through forums, Discord channels, or developer documentation. Other developers and community members may have encountered similar issues and can offer valuable insights or assistance.

While ZKSync lacks a dedicated transaction debugging tool like Tenderly, leveraging these approaches can help you diagnose and resolve transaction failures more effectively.

**Comment by clevelandmarcus** [comment]
Unfortunately, there isn't a direct equivalent to Tenderly for zkSync transactions. However, you can utilize the zkSync Explorer to inspect failed transactions, review logs, and gather insights into transaction failures. Additionally, implementing robust error handling and logging within your dapp can provide valuable information for debugging failed transactions.

**Comment by golfdutoresste** [comment]
For better understanding of transaction failures on zkSync, consider using zkSync Explorer which provides detailed transaction information including error messages. Additionally, leverage logging within your smart contracts to capture specific error conditions and debug transaction failures effectively.

**Comment by jennifercerso** [comment]
On zkSync, you can use the ethers.js library to get more information about transaction failures. Specifically, you can use the ethers.provider.getTransactionReceipt() method to get the transaction receipt, which contains the status field indicating whether the transaction succeeded or failed. Additionally, you can use the ethers.utils.parseTransaction() method to parse the transaction data and debug any issues. As for tools, zkSync provides a block explorer that can be used to view transaction details and logs.

**Comment by caojeremyv** [comment]
The user is building a dapp where transactions sometimes fail, and they are looking for tools to better understand why this is happening on zkSync. They are specifically asking if there is a stack trace or something similar that they can look at. The user previously used Tenderly on Ethereum but is unsure of any similar tools for zkSync.
